bots/github-push

```diff
@@ -0,0 +1,154 @@
+#! /usr/bin/python3
+
+# github-push -- Push the local branch to a GitHub PR, with benefits
+#
+# This tool pushes your local changes to a PR and also adds a comment
+# that summarizes what has changed.  The idea is to cement the actual
+# history of the PR in its comments, regardless of rebases, rewrites,
+# squashed fixups, etc.
+#
+# If you don't have a upstream tracking branch, you need to specify
+# the pull request number as an argument.
+#
+# If you do have a upstream tracking branch, you can omit the pull
+# request number.  The tool will find it by looking through all open
+# pull requests on GitHub and picking the one that matches the
+# tracking branch.  Once it has been found, the number is cached in
+# the Git config file to speed up subsequent invocations.
+
+import subprocess
+import argparse
+import sys
+import time
+
+from task import github
+
+def execute(*args):
+    try:
+        output = subprocess.check_output(args, universal_newlines=True)
+    except subprocess.CalledProcessError as ex:
+        sys.exit(ex.returncode)
+    return output
+
+def git(*args):
+    return execute("git", *args)
+
+def git_config(name):
+    try:
+        output = subprocess.check_output([ "git", "config", name ], universal_newlines=True)
+    except subprocess.CalledProcessError as ex:
+        return None
+    return output.strip()
+
+def push_and_comment(api, repo, pull_number, remote, local, push_dest, push_ref):
+    if local == remote:
+        sys.stderr.write("Nothing to push.\n")
+        return 0
+
+    diff = git("diff", remote, local)
+    old_commits_link = "https://github.com/" + repo + "/commits/" + remote
+    diff_link = "https://github.com/" + repo + "/compare/" + remote + ".." + local
+
+    def make_comment(title, include_old_commits=False):
+        diff_lines = diff.count("\n")
+        first_line_parts = [ ]
+        if include_old_commits:
+            first_line_parts.append("[Old commits](%s)" % old_commits_link)
+        if diff_lines > 0:
+            first_line_parts.append("[Diff](%s)" % (diff_link))
+        first_line = " | ".join(first_line_parts)
+        if diff_lines > 0 and diff_lines < 500:
```

Comment:
IMHO 500 is quite an enormous comment; the /compare link should suffice for most use cases and doesn't clobber the comments then. But this is not a veto, we can try this and change this in the future to our liking.